docs/it/refactoring.md

```diff
@@ -7,3 +7,132 @@
+Poniamo il caso che abbiamo una parte del nostro software con del codice "legacy", ovvero scritto più di tre mesi fà, che funziona egregiamente, fa il suo ma, ci rendiamo conto che le performance non sono quelle che ci attendiamo.
+Fatta una rapida analisi, notiamo che, magari, ci sono dei cicli ripetuti, o parti di codice che potrebbero essere evitate se solo avessimo posto dei semafori a monte.
+Con del buon refactoring potremmo rivedere il codice, eliminare delle dipendenze inutili o utilizzare dei costrutti del linguaggio più adatti, con il risultato di migliorare le performance.
+E se non sono le performance a preoccuparci, ma la poltiglia di codice scritto, magari di fretta e sotto pressione, che non rispetta principi base, come la responsabilità singola? Anche in questo caso, possiamo rivedere il codice, renderlo più leggibile e comprensibile così da ottenere un altro bel risultato: eliminare debito tecnico.
+Ma, come detto, le occasioni non mancano.
+Potremmo cogliere l'occasione di una Code Review o l'aggiunta di una nuova feature. Prima di iniziarne lo sviluppo, sicuramente è buona pratica leggere il codice in cui questa nuova funzionalità andrà a integrarsi e, se quest'ultimo risulta di difficile comprensione, è un'ottima occasione per effettuarne la rifattorizzazione. Questo ci permetterebbe di ottenere una maggiore conoscienza del codice.
+Un'altra, forse la migliore, dove è (quasi) obbligatorio, è la risoluzione di un bug in una parte di codice convulso, la cui sola lettura ci fa venire il mal di testa. Un'attività di refactoring potrebbe facilmente portare alla luce l'errore presente nel codice.
+L'attività di refactoring può essere paragonata a una pulizia profonda di casa, piuttosto che a una routine di sistemazione della propria cameretta, necessaria per mantenere il codice sicuro, di valore ed efficiente.
+
+## Quando è meglio evitare
+
+Non sempre è utile effettuare refactoring. Ci sono casi in cui il codice è già perfettamente ottimizzato anche se, a una vista inesperta, potrebbe risultare di difficile lettura.
+Ci sono casi in cui le performance sono fondamentali e i principi solid non sono sempre rispettati. In questi casi, il refactoring ci porterebbe a creare dei layer aggiuntivi di codice che, sebbene ci renderebbe il codice più comprensibile, avrebbe un impatto negativo sulle performance. Un esempio sono le librerie che permettono le connessioni con delle basi di dati, dove magari è fondamentale la gestione della latenza.
+
+## Come
+
+Come si può intuire, con le attività di refactoring andremo a migliorare il codice in modo da renderlo meno complesso, più mantenibile, più resiliente.
+Per farlo, potremmo iniziare con:
+
+- Red-Green-Refactor;
+- Rinominare i metodi oscuri;
+- Cambiare la firma dei metodi;
+- Estrarre nuovi metodi;
+- Semplificare i metodi esistenti;
+- Astrarre.
+
+### Red-Green-Refactor
+
+Forse il metodo più popolare, dove lo sviluppatore suddivide il processo di scrittura del codice in tre fasi:
```

Comment:
Indicherei che questo principio deriva dal TDD in qualche modo

Reply:
Aggiunto il riferimento al TDD dove, operativamente, si applica il Red-Green-Refactor